Investigate why trip tracking is glitching by adding additional logs

louisg1337 opened the pull request:
Add logs to the FSM to figure out why trips don't track or why their tracking is messed up on Android.

shankari commented:
@louisg1337 I will push this to staging once the UI changes have been pushed to production. you should also be able to manually kill the foreground service using `adb`, at least for testing why the foreground service doesn't restart after it is killed

louisg1337 commented:
**Previous Changes**

- Added in foreground service check in `ActivityRecognitionChangeIntentService` and checked to see if location services were available


**New Changes**

- Foreground service check when `DataCollectionPlugin.java / onResume()` gets called already implemented ([here](https://github.com/e-mission/e-mission-data-collection/blob/24d9aa58f141af39631890983a05dee18145300b/src/android/DataCollectionPlugin.java#L68))
- Added in log for when `DataCollectionPlugin.java` calls `onStop()`, `onDestroy()`, and `onSaveInstanceState()`
  - Last one is interesting, apparently it gets called when the plugin is communicating with an Activity that is getting killed by the OS ([docs here](https://github.com/apache/cordova-android/blob/3503bfa31b10cc7d6948356d4c101c7b76109688/framework/src/org/apache/cordova/CordovaPlugin.java#L189)).
- Check for FSM state in `ActivityRecognitionChangeIntentService`

shankari commented on src/android/DataCollectionPlugin.java:
```diff
@@ -232,4 +233,28 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
+     */
+    @Override
+    public Bundle onSaveInstanceState() {
+      Log.d(cordova.getActivity(), TAG, "onSaveInstanceState() has been called, OS killed an activity!");
```
nit: shouldn't this be "OS killed *this* activity"?

louisg1337 replied:
Yes I think you're right, I misread the docs sorry about that.

shankari commented on src/android/location/ActivityRecognitionChangeIntentService.java:
```diff
@@ -50,7 +56,38 @@ protected void onHandleIntent(Intent intent) {
+		}
+
+		// Add in logs to check for trip dissapearance
+		Log.d(this, TAG, "Beginning checks for trip dissapearance...");
```
nit: spelling

shankari replied:
nit: spelling

shankari commented on src/android/location/ActivityRecognitionChangeIntentService.java:
```diff
@@ -50,7 +56,38 @@ protected void onHandleIntent(Intent intent) {
+		});
+		
+		// Check for current state
+		SharedPreferences mPrefs = PreferenceManager.getDefaultSharedPreferences(this);
+        String mCurrState = mPrefs.getString(this.getString(R.string.curr_state_key),
+            this.getString(R.string.state_start));
+
+		Log.d(this, TAG, "Ending checks for trip dissapearance...");
```
we get this state, but we don't log it anywhere?!

louisg1337 replied:
sorry about that, thank you for catching that!

shankari commented on src/android/location/ActivityRecognitionChangeIntentService.java:
```diff
@@ -50,7 +56,38 @@ protected void onHandleIntent(Intent intent) {
+		Context ctxt = getApplicationContext();
+		TripDiaryStateMachineForegroundService.checkForegroundNotification(ctxt);
+
+		// Check if can retrieve current location
+		LocationServices.getFusedLocationProviderClient(ctxt).getLocationAvailability()
+		.addOnSuccessListener(locationAvailability -> {
+			if (locationAvailability != null && locationAvailability.isLocationAvailable()) {
+				// Location is available
+				Log.d(this, TAG, "Location is available");
+			} else {
+				// Location is not available
+				Log.d(this, TAG, "Location is not available");
 			}
+		})
+		.addOnFailureListener(e -> {
+			// Handle failure, if any
+			Log.d(this, TAG, "Error getting location availability: " + e.getMessage());
+		});
```
Instead of checking location availability directly, why not use the methods in the background checks?

louisg1337 replied:
Great point, I think I wrote this part a while ago when I wasn't too familiar with the code base. Converted this to use `checkLocationSettings` in `SensorControlChecks.java`.

shankari reviewed (commented):
I am generally supportive of this change, but would like to see the answers ASAP (if possible) so I can merge and push out a release to staging...

shankari commented on src/android/location/ActivityRecognitionChangeIntentService.java:
```diff
@@ -50,7 +60,30 @@ protected void onHandleIntent(Intent intent) {
+		// Check if can retrieve current location
+		OnCompleteListener<LocationSettingsResponse> callback = new OnCompleteListener<LocationSettingsResponse>() {
+			@Override // no-op
+			public void onComplete(Task<LocationSettingsResponse> task) { }
+		};
```
```suggestion
			public void onComplete(Task<LocationSettingsResponse> task) {
                           Log.d(this, TAG, "While checking location status in activity response, received callback with : " + task);
			}
		};
```

shankari commented:
@louisg1337 you also need to bump up the version in `plugin.xml`
